### DISCUSSION TITLE
Payment table in DB

### DISCUSSION THREAD

**Discussion by tiagonmas** [banco-alimentar/alimentestaideia.pt#192] [Q&A]
Hi @guerrerotook 
I noticed you moved away from different tables for each payment type and consolidated all in one table.
My concern is that people can try multiple payment types, and they could even try 2 times the same type, and if we use only that one table, we might lose information !?

Currently I can go back to invoice history and try to pay from there. If I initially selected multibanco (a reference was already created for me), but then after a couple of days I go back to that donation history and try to pay again, a new multibanco reference will be created and the history of the privious one is lost !?

Do you think it is an issue ? Possible solutions ?
For multibanco, we should show the previous reference (if it is still active) instead of creating a new one ?

For the other payment types I do worry that there might be several payment attempts and this way we don't have "history".

FYI @joalmeid

**Comment by guerrerotook** [comment]
Well, having all payment in a table was easy to manage from the database perspective and performance. I did this change because is going to be very difficult to manage 4 tables. Instead what we have all the information in the same table but using different columns. We use a technology in entity framework core called table inheritance, that let us have a column discriminator to know what is the inherited class in C# that we should use. 

Given said that, that doesn't mean that we should support having multiple ways for paying a donation. For now a donation can just have a Donation, as how it designed. But we can change that property to be a collection. That will let us have a collection of payments for each donation. If this is what you want I can make the change. Just let me know.

**Reply by tiagonmas** [reply]
I'm just worried of losing information, namely of our exchanges with easypay, if people try to pay the same donation mutiple times using the same method. IF we crush the previous information, it will be hard to "debug" things in the future, if we have no trace of those transactions with easypay. IMO, every "exchange" with easypay should be recorded.

the user might go back and forth with easypay trying to pay the same donation. Especially now that we have the "complete donation" in the history, since it starts a new payment again.

**Reply by joalmeid** [reply]
One way to consider would be to have a async process to log all payments. Was thinking of something SQL based, having triggers to copy all payments to a side table, even when a payment is updated (user tried different payment method).

This could also be implemented in code. Side effect would be a larger payment data stored (almost duplicated)

Although Easypay have a full log of all payments, it makes sense to isolate payment data @ BA.
